rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/monitoring/views/CompactFlowPanel.java

```diff
@@ -0,0 +1,426 @@
+                ++index;
+                index += hierarchyNodeSource.getDeploying().size();
+
+                int indexOfFirstHost = index;
+
+                for (HierarchyHost hierarchyHost : hierarchyNodeSource.getHosts()) {
+                    if (hierarchyHost.getHost().getHostName().equals(nodeTile.getNode().getHostName())) {
+                        ++index;
+
+                        hierarchyNodeSource.incrementTiles();
+                        hierarchyHost.incrementTiles();
+
+                        hierarchyHost.getNodes().add(0, nodeTile.getNode());
+
+                        this.insert(nodeTile, index);
+                        return;
+                    } else {
+                        index += hierarchyHost.getTiles();
+                    }
+                }
+
+                // if we are here than there is no host for this node yet
+                // thus we will add it and node after
+                hierarchyNodeSource.incrementTiles();
+                final HierarchyHost hierarchyHost = new HierarchyHost(hostTile.getHost());
+                hierarchyNodeSource.getHosts().add(0, hierarchyHost);
+                this.insert(hostTile, indexOfFirstHost);
+
+                hierarchyNodeSource.incrementTiles();
+                hierarchyHost.incrementTiles();
+                hierarchyHost.getNodes().add(0, nodeTile.getNode());
+                this.insert(nodeTile, indexOfFirstHost + 1);
+
+                return;
+
+            } else {
+                index += hierarchyNodeSource.getTiles();
+            }
+        }
+    }
+
+    protected void drawDeployingNode(CompactView.Tile nodeTile) {
+        int index = 0;
+        for (HierarchyNodeSource hierarchyNodeSource : model) {
+            if (hierarchyNodeSource.getNodeSource().getSourceName().equals(nodeTile.getNode().getSourceName())) {
+                ++index;
+
+                hierarchyNodeSource.incrementTiles();
+
+                hierarchyNodeSource.getDeploying().add(0, nodeTile.getNode());
+
+                this.insert(nodeTile, index);
+                return;
+            } else {
+                index += hierarchyNodeSource.getTiles();
+            }
+        }
+    }
+
+    public void remove(NodeSource nodeSource) {
+        int index = 0;
+        final Iterator<HierarchyNodeSource> iterator = model.iterator();
+        while (iterator.hasNext()) {
+            final HierarchyNodeSource hierarchyNodeSource = iterator.next();
+            if (hierarchyNodeSource.getNodeSource().equals(nodeSource)) {
+                for (int i = 0; i < hierarchyNodeSource.getTiles(); ++i) {
+                    remove(index);
+                }
+                iterator.remove();
+                return;
+            } else {
+                index += hierarchyNodeSource.getTiles();
+            }
+        }
+    }
+
+    public void remove(NodeSource.Host.Node node) {
```

Comment:
This method seems quite long and complex. Maybe the `NodeSource`/`Host`/`Node` classes can store more information so that we need to compute less things here?

Reply:
I agree that it is complex, i will refactor and split it. However, I do not know how to "store more info".